pkg/forman/forman_test.go

```diff
@@ -0,0 +1,119 @@
+package forman
+
+import (
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"go.uber.org/zap"
+)
+
+func TestGoForman(t *testing.T) {
+	workerCount := 2
+	// totalRequest is the total number of requests to be processed to delay work
+	totalRequest := 10
+	// Create a counter to track the number of requests processed
+	requestCounter := sync.Map{}
+	workWithDelay := "test-work-with-delay"
+	requeueWork := "test-work-requeue"
+	panicWork := "test-panic"
+
+	// Create a callback function to track the processing of requests
+	callback := func(req Request) {
+		val, ok := requestCounter.Load(req.Name)
+		if ok {
+			requestCounter.Store(req.Name, val.(int)+1)
+			return
+		}
+		requestCounter.Store(req.Name, 0)
+	}
+	// 3 types of request
+	type work struct {
+		name          string
+		reconcileFunc func(Request) Result
+	}
+	works := []work{
+		{
+			name: workWithDelay,
+			reconcileFunc: func(req Request) Result {
+				callback(req)
+				sleepTime := totalRequest
+				// sleep for half the total request duration if the total
+				// request is odd number, increment the sleep time by 1 to
+				// process the request only twice
+				if sleepTime%2 != 0 {
+					sleepTime++
+				}
+				time.Sleep((time.Duration(sleepTime) / 2) * time.Second)
+				return Result{}
+			},
+		}, {
+			name: requeueWork,
+			reconcileFunc: func(req Request) Result {
+				callback(req)
+				return Result{Requeue: true, After: 1 * time.Second}
+			},
+		}, {
+			name: panicWork,
+			reconcileFunc: func(req Request) Result {
+				callback(req)
+				time.Sleep(1 * time.Second)
+				panic("Simulated panic")
+			},
+		},
+	}
+
+	// Create a mock reconcile function that returns a successful result
+	reconcile := func(req Request) Result {
+		for _, work := range works {
+			if req.Name == work.name {
+				return work.reconcileFunc(req)
+			}
+		}
+		return Result{}
+	}
+
+	// Create a channel to receive requests
+	requests := GoForman(zap.NewNop(), workerCount, reconcile)
+
+	requests <- Request{Name: panicWork}
+	requests <- Request{Name: requeueWork}
+	// Send multiple requests for the same work
+	for i := 0; i < totalRequest; i++ {
+		requests <- Request{Name: workWithDelay}
+		time.Sleep(1 * time.Second)
+	}
+
+	// Get the current value of the request counter
+	previousRequestCounter := sync.Map{}
+	requestCounter.Range(func(k, v interface{}) bool {
+		previousRequestCounter.Store(k, v)
+		return true
+	})
```

Comment:
1. Yes it might have changed, this is a check to ensure that the count values are changed after the sleep.
2. This is just to avoid multiple Load statements, rather copy from the map and this will not get changed even if we add different types of work in the future.